pkg/spanconfig/spanconfigsqlwatcher/sql_watcher.go

```diff
@@ -0,0 +1,453 @@
+					ev.Key,
+				)
+				return
+			}
+			descID = descpb.ID(tree.MustBeDInt(startKeyRow[0].Datum))
+		}
+
+		select {
+		case <-ctx.Done():
+		case updatesCh <- descID:
+		}
+	}
+	rf, err := s.rangeFeedFactory.RangeFeed(
+		ctx,
+		"sql-watcher-zones-rangefeed",
+		descriptorTableSpan,
+		s.clock.Now(),
+		handleEvent,
+	)
+	if err != nil {
+		return nil, err
+	}
+	s.stopper.AddCloser(rf)
+
+	log.Infof(ctx, "established range feed over system.zones table")
+	return updatesCh, nil
+}
+
+// getAllAffectedDescriptors returns a list of IDs that need to have their span
+// configurations refreshed because the given id changed.
+func getAllAffectedIDs(
+	ctx context.Context, id descpb.ID, txn *kv.Txn, descsCol *descs.Collection,
+) (descpb.IDs, error) {
+	desc, err := descsCol.GetImmutableDescriptorByID(ctx, txn, id, tree.CommonLookupFlags{
+		IncludeDropped: true,
+	})
+	if err != nil {
+		if errors.Is(err, catalog.ErrDescriptorNotFound) {
+			return descpb.IDs{id}, nil
+		}
+		return nil, err
+	}
+	switch desc.DescriptorType() {
+	case catalog.Table:
+		return descpb.IDs{id}, nil
+	case catalog.Database:
+	default:
+		return nil, errors.AssertionFailedf("expected either database or table, but found descriptor of type [%s]", desc.DescriptorType())
+	}
+	// There's nothing to do here if the database has been dropped. If the
+	// database was non-empty those objects will have their own rangefeed events.
+	if desc.Dropped() {
+		return nil, nil
+	}
+	descriptors, err := descsCol.GetAllDescriptorsInDatabase(ctx, txn, id)
+	if err != nil {
+		return nil, err
+	}
+	ret := make(descpb.IDs, 0, len(descriptors))
+	for _, desc := range descriptors {
+		ret = append(ret, desc.GetID())
+	}
+	return ret, nil
+}
+
+// onDescIDUpdate generates the span configurations for the given descriptor ID
+// and sends them to the provided updatesCh.
+func (s *SQLWatcher) onDescIDUpdate(
+	ctx context.Context, descID descpb.ID, updatesCh chan spanconfig.Update,
+) error {
+	var updates []spanconfig.Update
+	if err := descs.Txn(
+		ctx,
+		s.settings,
+		s.leaseManager,
+		s.ie,
+		s.DB,
+		func(ctx context.Context, txn *kv.Txn, descsCol *descs.Collection) error {
+			affectedIDs, err := getAllAffectedIDs(ctx, descID, txn, descsCol)
+			if err != nil {
+				return err
+			}
+			entries, err := s.generateSpanConfigurations(ctx, txn, affectedIDs)
+			if err != nil {
+				return err
+			}
+			for _, entry := range entries {
+				update := spanconfig.Update{
+					Entry:   entry,
+					Deleted: false,
+				}
+				{
+					// Try to get the table descriptor regardless of DROP status. The span
+					// config entry only needs to be removed when the descriptor is
+					// deleted, not when it is dropped, so we set the Deleted flag only if
+					// no descriptor with the given ID exists.
+					_, err := descsCol.GetImmutableDescriptorByID(ctx, txn, descID, tree.CommonLookupFlags{
+						AvoidCached:    true,
+						IncludeDropped: true,
+					})
+					if err != nil {
```

Comment:
Fewer levels of indent if you check for ErrDescriptorNotFound first, err != nil in the else block.